### PR TITLE
[bug](fold) fix fold constant core dump with variant type

### PR DESCRIPTION
**Pull request by zhangstar333**
## Proposed changes
1. variant type core dump at call get_data_at function, as not impl this function.
2. some case can't pass at old planner and  fold_constant_by_be = on.
3. open enable_fold_constant_by_be = true.
```
F20240314 19:24:30.485731 1887808 column_object.h:415] should not call the method in column object
*** Check failure stack trace: ***
    @     0x55a27aeb85f6  google::LogMessage::SendToLog()
    @     0x55a27aeb5040  google::LogMessage::Flush()
    @     0x55a27aeb8e39  google::LogMessageFatal::~LogMessageFatal()
    @     0x55a2744d0521  doris::vectorized::ColumnObject::get_data_at()
    @     0x55a2744b9f28  doris::vectorized::ColumnNullable::get_data_at()
    @     0x55a2716979b8  doris::FoldConstantExecutor::fold_constant_vexpr()
    @     0x55a271661126  doris::PInternalServiceImpl::_fold_constant_expr()
    @     0x55a27166e01d  std::_Function_handler<>::_M_invoke()
    @     0x55a27167d9cb  doris::WorkThreadPool<>::work_thread()
    @     0x55a27dc23050  execute_native_thread_routine
    @     0x7f01f659c609  start_thread
    @     0x7f01f6849133  clone
    @              (nil)  (unknown)
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

File: regression-test/suites/nereids_tpcds_shape_sf100_p0/rf_prune/query21.groovy [hunk 1]
```diff
@@ -22,6 +22,7 @@ suite("query21") {
     sql "use ${db}"
     sql 'set enable_nereids_planner=true'
     sql 'set enable_fallback_to_original_planner=false'
+    sql 'SET enable_fold_constant_by_be = false' //plan shape will be different
```
Comment (HappenLee) [inline regression-test/suites/nereids_tpcds_shape_sf100_p0/rf_prune/query21.groovy:25]: better change the plan

File: regression-test/suites/datatype_p0/agg_state/nereids/test_agg_state_nereids.groovy [hunk 2]
```diff
@@ -16,9 +16,10 @@
 // under the License.
 
 suite("test_agg_state_nereids") {
-    sql "set global enable_agg_state=true"
+    sql "set enable_agg_state=true"
     sql "set enable_nereids_planner=true;"
     sql "set enable_fallback_to_original_planner=false;"
+    sql "set enable_fold_constant_by_be=false;"
```
Comment (HappenLee) [inline regression-test/suites/datatype_p0/agg_state/nereids/test_agg_state_nereids.groovy:22]: check the case

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by BiteTheDDDDt — APPROVED** [review]

**Pull request merged**